Readability.js

```diff
@@ -1798,6 +1807,17 @@ Readability.prototype = {
     return weight;
   },
 
+  _getTotalClassWeight: function(node) {
+    let weight = this._getClassWeight(node);
+
+    let children = node.getElementsByTagName("*");
+    for (const child of children) {
```

Comment:
@gijsk I'm not sure why we compute the weight of innermost child 5 times:
```js
for (const child of children) {
  weight += this._getClassWeight(child);
}
```
I'm not calling recursive here, the `_getClassWeight` function is still the same and only calculate the weight for that specific node and not its children or descendants. Can you help clarify?

Reply:
@thomasdao so every time that code runs for a given node, it evaluates the class weight of that node and all its descendants,  adding all of them together.

But `_cleanConditionally` gets called a number of times, with different tag names. In the example with the nested divs you gave in the main comment, the call at https://github.com/mozilla/readability/blob/e1972362816cd47e679d1a4bbbf923a58b8a88d5/Readability.js#L706 will run this code for every `div`, so once for the parent div, then again for the child div -- but if there are more `div` elements higher up in the document tree (which is likely!), it'll run this for all of those as well.

So the fact that the code you're adding runs `getElementsByTagName("*")`, collecting all the descendants, for every node on which the inner callback of `_cleanConditionally` is called, is likely what causes us to evaluate nodes more than once.

To fix this, I would suggest that rather than putting this in `_cleanConditionally`, instead collect all `img` elements in the document, loop over those and for all the large ones, walk up their parent tree and store the "image bonus" on the ancestors directly. You could add another "private" (`_`-prefixed) method for this and call it after we unwrap no script images, here: https://github.com/mozilla/readability/blob/e1972362816cd47e679d1a4bbbf923a58b8a88d5/Readability.js#L2212

This also avoids the other issue here, which is how you weight the class weight of the descendants when applying it to ancestors (for which we otherwise use a multiplier factor, https://github.com/mozilla/readability/blob/e1972362816cd47e679d1a4bbbf923a58b8a88d5/Readability.js#L1036-L1046 ).